pkg/kds/remote/components.go

```diff
@@ -41,34 +39,34 @@ var (
 	}
 )
 
-func SetupServer(rt core_runtime.Runtime) error {
-	hasher, cache := kds_server.NewXdsContext(kdsRemoteLog)
-	generator := kds_server.NewSnapshotGenerator(rt, providedTypes, providedFilter(rt.Config().Mode.Remote.Zone))
-	versioner := kds_server.NewVersioner()
-	reconciler := kds_server.NewReconciler(hasher, cache, generator, versioner)
-	syncTracker := kds_server.NewSyncTracker(kdsRemoteLog, reconciler, rt.Config().KDS.Server.RefreshInterval)
-	resourceSyncer := sync_store.NewResourceSyncer(kdsRemoteLog, rt.ResourceStore())
-
-	clientFactory := func(clusterAddress string) kds_client.ClientFactory {
-		return func() (kdsClient kds_client.KDSClient, err error) {
-			return kds_client.New(clusterAddress, rt.Config().KDS.Client)
-		}
-	}
-
-	componentFactory := func(log logr.Logger, req *envoy_api_v2.DiscoveryRequest) component.Component {
-		globalAddress := req.Node.Id
-		policiesSink := kds_client.NewKDSSink(kdsRemoteLog, rt.Config().Mode.Remote.Zone, consumedTypes,
-			clientFactory(globalAddress), Callbacks(resourceSyncer, rt.Config().Store.Type == store.KubernetesStore, rt.Config().Mode.Remote.Zone))
-		return component.NewResilientComponent(kdsRemoteLog, policiesSink)
-	}
-
-	callbacks := util_xds.CallbacksChain{
-		util_xds.LoggingCallbacks{Log: kdsRemoteLog},
-		syncTracker,
-		NewComponentSpawner(kdsRemoteLog.WithName("policy-sink-spawner"), componentFactory),
+func Setup(rt core_runtime.Runtime) error {
+	kdsServer, err := kds_server.New(kdsRemoteLog, rt, providedTypes, rt.Config().Mode.Remote.Zone, providedFilter(rt.Config().Mode.Remote.Zone))
+	if err != nil {
+		return err
 	}
-	srv := kds_server.NewServer(cache, callbacks, kdsRemoteLog, rt.Config().Mode.Remote.Zone)
-	return rt.Add(kds_server.NewKDSServer(srv, *rt.Config().KDS.Server))
+	resourceSyncer := sync_store.NewResourceSyncer(kdsRemoteLog, rt.ResourceStore())
+	onSessionStarted := mux.OnSessionStartedFunc(func(session mux.Session) error {
+		kdsRemoteLog.Info("new session created", "peerID", session.PeerID())
+		go func() {
+			if err := kdsServer.StreamKumaResources(session.ServerStream()); err != nil {
+				kdsRemoteLog.Error(err, "KDSSink finished with an error")
+			}
+		}()
+		sink := kds_client.NewKDSSink(
+			kdsRemoteLog.WithValues("zone", session.PeerID()),
+			consumedTypes,
+			kds_client.NewKDSStream(session.ClientStream(), rt.Config().Mode.Remote.Zone),
+			Callbacks(resourceSyncer, rt.Config().Store.Type == store.KubernetesStore, rt.Config().Mode.Remote.Zone),
+		)
+		go func() {
+			if err := sink.Start(session.Done()); err != nil {
+				kdsRemoteLog.Error(err, "KDSSink finished with an error")
+			}
+		}()
+		return nil
+	})
+	muxClient := mux.NewClient(rt.Config().KDS.Client.GlobalAddress, rt.Config().Mode.Remote.Zone, onSessionStarted, *rt.Config().KDS.Client)
```

Comment:
Will it make sense to have something like: `remoteZone := rt.Config().Mode.Remote.Zone` and then use `remoteZone` throughout the function instead?